readme.md
# BloomSky Sky1/Sky2/Storm Driver #

## Description ##

The BloomSky driver is a WeeWX driver that supports the BloomSky Sky1, Sky2 and Storm personal weather stations. The driver utilises the BloomSky API to obtain observation data from the BloomSky devices.

## Pre-Requisites ##

The BloomSky driver requires WeeWX v3.7.0 or greater. Both Python 2 and Python 3 are supported when using WeeWX v4.0.0 or later. A BloomSky API key is also required and can be obtained from [dashboard.bloomsky.com](dashboard.bloomsky.com).

## Installation Instructions ##

### Installation using the *wee_extension* utility ###

**Note**:   Symbolic names are used below to refer to some file location on the WeeWX system. These symbolic names allow a common name to be used to refer to a directory that may be different from system to system. The following symbolic names are used below:

-   *$DOWNLOAD_ROOT*. The path to the directory containing the downloaded BloomSky driver extension.

-   *$BIN_ROOT*. The path to the directory where WeeWX executables are located. This directory varies depending on WeeWX installation method. Refer to [where to find things](http://weewx.com/docs/usersguide.htm#Where_to_find_things) in the WeeWX User's Guide for further information.

1.  Download the latest BloomSky driver extension from the [Bloomsky driver releases page](https://github.com/gjr80/weewx-bloomsky/releases) into a directory accessible from the WeeWX machine.

        $ wget -P $DOWNLOAD_ROOT https://github.com/gjr80/weewx-bloomsky/releases/download/v2.0.2/bloomsky-2.0.2.tar.gz

    where *$DOWNLOAD_ROOT* is the path to the directory where the BloomSky driver extension is to be downloaded.

1.  Stop WeeWX:

        $ sudo /etc/init.d/weewx stop

    or

        $ sudo service weewx stop
        
    or
    
        $ sudo systemctl stop weewx

1.  Install the BloomSky driver extension downloaded at step 1 using the *wee_extension* utility:

        $ wee_extension --install=$DOWNLOAD_ROOT/bloomsky-2.0.2.tar.gz

    This will result in output similar to the following:

        Request to install '/var/tmp/bloomsky-2.0.2.tar.gz'
        Extracting from tar archive /var/tmp/bloomsky-2.0.2.tar.gz
        Saving installer file to /home/weewx/bin/user/installer/Bloomsky
        Saved configuration dictionary. Backup copy at /home/weewx/weewx.conf.20200523124410
        Finished installing extension '/var/tmp/bloomsky-2.0.2.tar.gz'

1.  Select and configure the driver:

        $ sudo wee_config --reconfigure

    selecting *BloomSky (user.bloomsky)* and providing the BloomSky API key to be used when prompted.
    
    **Note**: If you do not yet have a BloomSky API key enter a fictitious API key and later when you do have an API key edit weewx.conf and set the *[Bloomsky] api_key* option appropriately.

1.  Add the following stanza to *weewx.conf*:

        [Accumulator]
            [[deviceID]]
                adder = noop
            [[deviceName]]
                adder = noop
            [[imageURL]]
                adder = noop
            [[deviceType]]
                adder = noop
            [[night]]
                adder = noop
            [[imageTimestamp]]
                adder = noop

1.  Start WeeWX:

        $ sudo /etc/init.d/weewx start

    or

        $ sudo service weewx start
        
    or
    
        $ sudo systemctl start weewx

This will result in the driver collecting data from the bloomsky API and emitting loop packets. The WeeWX log may be monitored to confirm operation. The BloomSky driver installation can be further customized (eg sensor mapping, polling interval etc) by referring to the BloomSky driver extension wiki or the comments at the start of the BloomSky driver file *$BIN_ROOT/user/bloomsky.py*.

### Manual installation ###

1.  Download the latest BloomSky driver extension from the [Bloomsky driver releases page](https://github.com/gjr80/weewx-bloomsky/releases) into a directory accessible from the WeeWX machine.

        $ wget -P $DOWNLOAD_ROOT https://github.com/gjr80/weewx-bloomsky/releases/download/v2.0.2/bloomsky-2.0.2.tar.gz

    where *$DOWNLOAD_ROOT* is the path to the directory where the BloomSky driver extension is to be downloaded.

1.  Unpack the extension as follows:

        $ tar xvfz bloomsky-2.0.2.tar.gz

1.  Copy the *bloomsky.py* file from within the resulting folder as follows:

        $ cp bloomsky/bin/user/bloomsky.py $BIN_ROOT/user

    replacing the symbolic name *$BIN_ROOT* with the nominal location for your installation.

1.  Edit *weewx.conf*:

        $ nano weewx.conf

1.  Make the following changes to *weewx.conf*:

    - change the *[Station] station_type* config option to:

            [Station]
                ...
                station_type = Bloomsky

    -   add a *[Bloomsky]* stanza as follows:

            [Bloomsky]
                # This section is for the BloomSky station
    
                # BloomSky API key obtained from dashboard.bloomsky.com
                api_key = INSERT_API_KEY_HERE
    
                # BloomSky claim data is updated from the station every 5-8 minutes.
                # How often in seconds the driver will poll the BloomSky API. Default is
                # 60 seconds
                poll_interval = 60
    
                # The driver itself
                driver = user.bloomsky

    -   if it does not already exist add an *[Accumulator]* stanza as follows:

            [Accumulator]
                [[deviceID]]
                    adder = noop
                [[deviceName]]
                    adder = noop
                [[imageURL]]
                    adder = noop
                [[deviceType]]
                    adder = noop
                [[night]]
                    adder = noop
                [[imageTimestamp]]
                    adder = noop

        if the *[Accumulator]* stanza already exists ensure the child config options above are added.

1.  Start WeeWX:

        $ sudo /etc/init.d/weewx start

    or

        $ sudo service weewx start
        
    or
    
        $ sudo systemctl start weewx

This will result in the driver collecting data from the bloomsky API and emitting loop packets. The WeeWX log may be monitored to confirm operation. The BloomSky driver installation can be further customized (eg sensor mapping, polling interval etc) by referring to the BloomSky driver extension wiki or the comments at the start of the BloomSky driver file *$BIN_ROOT/user/bloomsky.py*.


changelog
v2.0.2
*   removed premature logging call
v2.0.1
*   fixed incorrect barometer/pressure field name in default field map
v2.0.0
*   WeeWX 3.7+/4.x python2/3 compatible
*   restructured main()
*   restructured options menu when driver is run directly
*   accumulator 'noop' adder now added for field 'raining' on install
v1.0.0
*   now works under python 2.6+ or python 3.5+ (python 2.5 no longer
    supported - use v0.1.1)
v0.2.0rc2
*   added ability to support multiple device IDs
*   now supports user defined sensor map
*   default poll interval now 60 seconds not 15 seconds
v0.1.1
*   added missing barometer/Pressure mapping to default sensor map
*   changed default poll interval to 60 seconds (as advertised)
*   revised driver comments
v0.1.0
*   initial release

install.py
"""
This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

                        Installer for Bloomsky Driver

Version: 2.0.2                                          Date: 6 September 2021

Revision History
    6 September 2021    v2.0.2
        - bump version number only
    8 June 2021         v2.0.1
        - bump version number only
    27 August 2020      v2.0.0
        - now WeeWX 3.7+/4.x python2/3 compatible
        - added accumulator noop adder for field raining
    31 May 2019         v1.0.0
        - bump version number only
    29 May 2019         v0.1.1
        - reformatted comments
    9 January 2018      v0.2.0rc2
        - added ability to support multiple device IDs
        - now supports user defined sensor map
        - default poll interval now 60 seconds not 15 seconds
    25 June 2017        v0.1.0
        - initial implementation
"""

import weewx

from distutils.version import StrictVersion
from setup import ExtensionInstaller

REQUIRED_VERSION = "3.7.0"
BLOOMSKY_VERSION = "2.0.2"


def loader():
    return BloomskyInstaller()


class BloomskyInstaller(ExtensionInstaller):
    def __init__(self):
        if StrictVersion(weewx.__version__) < StrictVersion(REQUIRED_VERSION):
            msg = "%s requires weeWX %s or greater, found %s" % (''.join(('Bloomsky driver ', BLOOMSKY_VERSION)),
                                                                 REQUIRED_VERSION,
                                                                 weewx.__version__)
            raise weewx.UnsupportedFeature(msg)
        super(BloomskyInstaller, self).__init__(
            version=BLOOMSKY_VERSION,
            name='Bloomsky',
            description='WeeWX driver for Bloomsky Sky1/Sky2/Storm personal weather stations.',
            author="Gary Roderick",
            author_email="gjroderick<@>gmail.com",
            files=[('bin/user', ['bin/user/bloomsky.py'])],
            config={
                'Bloomsky': {
                    'api_key': 'INSERT_API_KEY_HERE',
                    'poll_interval': 60,
                    'driver': 'user.bloomsky'
                },
                'Accumulator': {
                    'deviceID': {
                        'adder': 'noop'
                    },
                    'deviceName': {
                        'adder': 'noop'
                    },
                    'imageURL': {
                        'adder': 'noop'
                    },
                    'deviceType': {
                        'adder': 'noop'
                    },
                    'night': {
                        'adder': 'noop'
                    },
                    'imageTimestamp': {
                        'adder': 'noop'
                    },
                    'raining': {
                        'adder': 'noop'
                    }
                }
            }
        )
